pkg/auth/requests/service_account.go

```diff
@@ -0,0 +1,161 @@
+	// Basic checks to see if we even need to proceed
+	info, hasAuth := request.UserFrom(req.Context())
+	if info.GetName() != "system:cattle:error" {
+		return info, hasAuth, nil
+	}
+	clusterID := mux.Vars(req)["clusterID"]
+	if clusterID == "" {
+		return info, hasAuth, fmt.Errorf("no clusterID found in request")
+	}
+	// Check the cluster setting value to determine whether we will continue the auth process
+	settings, err := t.clusterProxyConfigsGetter.Get(clusterID, settingsObjectName)
+	if err != nil {
+		logrus.Warnf("rejecting downstream proxy request for %s, unable to fetch ClusterProxySettings object for cluster", req.URL.Path)
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	if !settings.Enabled {
+		logrus.Warnf("rejecting downstream proxy request for %s, current setting is enabled: %v", req.URL.Path, settings.Enabled)
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	// See if the token is a JWT.
+	rawToken := tokens.GetTokenAuthFromRequest(req)
+
+	jwtParser := jwtv4.Parser{}
+	claims := jwtv4.RegisteredClaims{}
+	_, _, err = jwtParser.ParseUnverified(rawToken, &claims)
+	if err != nil {
+		logrus.Debugf("saauth: error parsing JWT: %v", err)
+		return info, false, err
+	}
+
+	if !strings.HasPrefix(claims.Subject, serviceAccountSubjectPrefix) {
+		logrus.Debugf("saauth: JWT sub is not a service account: %v", err)
+		return info, false, nil
+	}
+
+	if isTokenExpired(claims.ExpiresAt) {
+		logrus.Debugf("saauth: Service Account JWT is expired. Expiration time was: %v", claims.ExpiresAt)
+		return info, false, nil
+	}
+
+	cluster, err := t.clusterLister.Get("", clusterID)
+	if err != nil {
+		logrus.Debugf("saauth: error getting cluster: %v", err)
+		return info, false, err
+	}
+
+	// Get rest config for the cluster and instantiate an authentication client.
+	kubeConfig, err := t.restConfigGetter(cluster, t.scaledContext, t.secretLister)
+	if kubeConfig == nil || err != nil {
+		return info, false, err
+	}
+
+	authClient, err := t.authClientCreator(kubeConfig)
+	if err != nil {
+		logrus.Debugf("saauth: error creating authentication client: %v", err)
+		return info, false, err
+	}
+
+	tokenReview := &v1.TokenReview{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "authentication.k8s.io/v1",
+			Kind:       "TokenReview",
+		},
+		Spec: v1.TokenReviewSpec{
+			Token: rawToken,
+		},
+	}
+
+	// Make the token review request to the downstream cluster.
+	tokenReview, err = authClient.TokenReviews().Create(req.Context(), tokenReview, metav1.CreateOptions{})
+	if err != nil {
+		logrus.Debugf("saauth: error creating a tokenreview request: %v", err)
+		return info, false, nil
+	}
+
+	// Let others know that this request is authenticated using a service account.
+	*req = *req.WithContext(authcontext.SetSAAuthenticated(req.Context()))
+
+	return &user.DefaultInfo{
+		Name:   tokenReview.Status.User.Username,
+		UID:    tokenReview.Status.User.UID,
+		Groups: tokenReview.Status.User.Groups,
```

Comment:
It would also be good to include the extras here.

Reply:
Yes.  Done.